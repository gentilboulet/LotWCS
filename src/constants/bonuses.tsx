export const BONUS_REDUCTION = 'BONUS_REDUCTION';
export type BONUS_REDUCTION = typeof BONUS_REDUCTION;

export const BONUS_DESTINY = 'BONUS_DESTINY';
export type BONUS_DESTINY = typeof BONUS_DESTINY;

export const BONUS_ENTANGLEMENT = 'BONUS_ENTANGLEMENT';
export type BONUS_ENTANGLEMENT = typeof BONUS_ENTANGLEMENT;

export const BONUS_STARTING_CHI = 'BONUS_STARTING_CHI';
export type BONUS_STARTING_CHI = typeof BONUS_STARTING_CHI;

export const BONUS_ONE_AMONG_N = 'BONUS_ONE_AMONG_N';
export type BONUS_ONE_AMONG_N = typeof BONUS_ONE_AMONG_N;

export const BONUS_SKILL_RANK = 'BONUS_SKILL_RANK';
export type BONUS_SKILL_RANK = typeof BONUS_SKILL_RANK;

export const BONUS_SPECIALITY = 'BONUS_SPECIALITY';
export type BONUS_SPECIALITY = typeof BONUS_SPECIALITY;
