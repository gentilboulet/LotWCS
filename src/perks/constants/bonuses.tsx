export const BONUS_DISCOUNT = 'BONUS_DISCOUNT';
export type BONUS_DISCOUNT = typeof BONUS_DISCOUNT;

export const BONUS_DESTINY = 'BONUS_DESTINY';
export type BONUS_DESTINY = typeof BONUS_DESTINY;

export const BONUS_ENTANGLEMENT = 'BONUS_ENTANGLEMENT';
export type BONUS_ENTANGLEMENT = typeof BONUS_ENTANGLEMENT;

export const BONUS_CHI = 'BONUS_CHI';
export type BONUS_CHI = typeof BONUS_CHI;

export const BONUS_CULTIVATION = 'BONUS_CULTIVATION';
export type BONUS_CULTIVATION = typeof BONUS_CULTIVATION;

export const BONUS_ONE_AMONG_N = 'BONUS_ONE_AMONG_N';
export type BONUS_ONE_AMONG_N = typeof BONUS_ONE_AMONG_N;

export const BONUS_SKILL_RANK = 'BONUS_SKILL_RANK';
export type BONUS_SKILL_RANK = typeof BONUS_SKILL_RANK;

export const BONUS_SPECIALITY = 'BONUS_SPECIALITY';
export type BONUS_SPECIALITY = typeof BONUS_SPECIALITY;
