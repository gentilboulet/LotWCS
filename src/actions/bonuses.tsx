import * as constants from '../constants/bonuses';
import { IBonus,
  IBonusDestiny, IBonusEntanglement, IBonusStartingChi, IBonusOneAmongN,
  IBonusSkillRank, IBonusSpeciality,
} from '../types/bonuses';

export function destiny(v: number): IBonusDestiny {
    return { type: constants.BONUS_DESTINY, value: v};
}

export function entanglement(v: number): IBonusEntanglement {
  return { type: constants.BONUS_ENTANGLEMENT, value: v};
}

export function startingChi(v: number): IBonusStartingChi {
  return { type: constants.BONUS_STARTING_CHI, value: v};
}

export function oneAmongN(bonuses: IBonus[]): IBonusOneAmongN {
  return { type: constants.BONUS_ONE_AMONG_N, bonuses: bonuses };
}

export function skillRank(skill: string): IBonusSkillRank {
  return { type: constants.BONUS_SKILL_RANK, skill: skill };
}

export function speciality(skill: string, specialityName: string): IBonusSpeciality {
  return { type: constants.BONUS_SPECIALITY, skill: skill, speciality: specialityName };
}
