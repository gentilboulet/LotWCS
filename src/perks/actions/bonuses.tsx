import * as constants from 'perks/constants/bonuses';
import { IBonus,
  IBonusDestiny, IBonusEntanglement, IBonusChi, IBonusCultivation, IBonusOneAmongN,
  IBonusSkillRank, IBonusSpeciality,
} from 'perks/types/bonuses';
import * as dataChi from 'data/chi';
import * as dataSkill from 'data/skills';

export function destiny(value: number): IBonusDestiny {
    return { type: constants.BONUS_DESTINY, value };
}

export function entanglement(value: number): IBonusEntanglement {
  return { type: constants.BONUS_ENTANGLEMENT, value };
}

export function chi(value: number, chiName: dataChi.IChiNames): IBonusChi {
  dataChi.validateChi(chiName);
  return { type: constants.BONUS_CHI, chi: chiName, value};
}

export function cultivation(value: number, cultivationName: dataChi.IChiCultivations): IBonusCultivation {
  dataChi.validateCultivation(cultivationName);
  return { type: constants.BONUS_CULTIVATION, value, cultivation: cultivationName};
}

export function oneAmongN(bonuses: IBonus[]): IBonusOneAmongN | IBonus {
  if (bonuses.length === 0) { throw new Error('Can not create one out of zero'); }
  if (bonuses.length === 1) { return bonuses[0]; }
  return { type: constants.BONUS_ONE_AMONG_N, bonuses };
}

export function skillRank(skill: string): IBonusSkillRank {
  dataSkill.validateSkill(skill);
  return { type: constants.BONUS_SKILL_RANK, skill };
}

export function speciality(skill: string, specialityName: string): IBonusSpeciality {
  dataSkill.validateSkill(skill);
  return { type: constants.BONUS_SPECIALITY, skill, speciality: specialityName };
}
