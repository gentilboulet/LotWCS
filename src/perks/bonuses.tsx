import { TChiName } from "../data/chi";
import { TSkillName } from "../data/skills";
import * as constants from "./constants/bonuses";

export interface IBonusDestiny {
  type: constants.BONUS_DESTINY;
  value: number;
}

export interface IBonusEntanglement {
  type: constants.BONUS_ENTANGLEMENT;
  value: number;
}

export interface IBonusChi {
  type: constants.BONUS_CHI;
  chi: TChiName;
  value: number;
}

export interface IBonusCultivation {
  type: constants.BONUS_CULTIVATION;
  chi: TChiName;
  value: number;
}

export interface IBonusOneAmongN {
  type: constants.BONUS_ONE_AMONG_N;
  bonuses: { [key in string]: IBonus };
}

// TODO : validity & auto resolution ?
export interface IBonusSkillRank {
  type: constants.BONUS_SKILL_RANK;
  skill: TSkillName;
}

export interface IBonusSpeciality {
  type: constants.BONUS_SPECIALITY;
  skill: TSkillName;
  speciality: string;
}

export type IBonus =
  | IBonusDestiny
  | IBonusEntanglement
  | IBonusChi
  | IBonusCultivation
  | IBonusOneAmongN
  | IBonusSkillRank
  | IBonusSpeciality;

export function isBonus(bonus: any): boolean {
  switch (bonus.type) {
    case constants.BONUS_DESTINY:
    case constants.BONUS_ENTANGLEMENT:
    case constants.BONUS_CHI:
    case constants.BONUS_CULTIVATION:
    case constants.BONUS_ONE_AMONG_N:
    case constants.BONUS_SKILL_RANK:
    case constants.BONUS_SPECIALITY:
      return true;
    case constants.BONUS_DISCOUNT:
    default:
      return false;
  }
}
