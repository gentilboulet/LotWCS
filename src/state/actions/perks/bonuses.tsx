import { IBonus,
    IBonusChi, IBonusCultivation, IBonusDestiny, IBonusEntanglement, IBonusOneAmongN,
    IBonusSkillRank, IBonusSpeciality,
} from 'state/bonuses';
import * as constants from 'state/constants/perks/bonuses';

import * as dataChi from 'data/chi';
import * as dataSkill from 'data/skills';

export function bonusDestiny(value: number): IBonusDestiny {
    return { type: constants.BONUS_DESTINY, value };
}

export function bonusEntanglement(value: number): IBonusEntanglement {
  return { type: constants.BONUS_ENTANGLEMENT, value };
}

export function bonusChi(value: number, chiName: dataChi.TChiName): IBonusChi {
  return { type: constants.BONUS_CHI, chi: chiName, value};
}

export function bonusCultivation(chiName: dataChi.TChiName, value: number): IBonusCultivation {
  return { type: constants.BONUS_CULTIVATION, value, chi: chiName};
}

export function bonusOneAmongN(bonuses: IBonus[]): IBonusOneAmongN | IBonus {
  if (bonuses.length === 0) { throw new Error('Can not create one out of zero'); }
  if (bonuses.length === 1) { return bonuses[0]; }
  return { type: constants.BONUS_ONE_AMONG_N, bonuses };
}

export function bonusSkillRank(skill: dataSkill.TSkillName): IBonusSkillRank {
  return { type: constants.BONUS_SKILL_RANK, skill };
}

export function bonusSpeciality(skill: dataSkill.TSkillName, specialityName: string): IBonusSpeciality {
  return { type: constants.BONUS_SPECIALITY, skill, speciality: specialityName };
}
