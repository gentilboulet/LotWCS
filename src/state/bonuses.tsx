import produce from "immer"

import * as chi from 'state/chi';
import { maxSkillBonus } from 'state/derived';
import * as skills from 'state/skills';

import { IStoreState } from 'state/type';

import { TChiName } from 'data/chi';
import * as constants from 'state/constants/perks/bonuses';

export interface IBonusDestiny {
  type: constants.BONUS_DESTINY;
  value: number;
}

export interface IBonusEntanglement {
  type: constants.BONUS_ENTANGLEMENT;
  value: number;
}

export interface IBonusChi {
  type: constants.BONUS_CHI;
  chi: TChiName;
  value: number;
}

export interface IBonusCultivation {
  type: constants.BONUS_CULTIVATION;
  chi: TChiName;
  value: number;
}

export interface IBonusOneAmongN {
  type: constants.BONUS_ONE_AMONG_N;
  bonuses: IBonus[];
}

export interface IBonusSkillRank {
  type: constants.BONUS_SKILL_RANK;
  skill: string;
}

export interface IBonusSpeciality {
  type: constants.BONUS_SPECIALITY;
  skill: string;
  speciality: string;
}

export type IBonus =
  IBonusDestiny
  | IBonusEntanglement
  | IBonusChi
  | IBonusCultivation
  | IBonusOneAmongN
  | IBonusSkillRank
  | IBonusSpeciality
;

export function isBonus(bonus: any): boolean {
  switch (bonus.type) {
    case constants.BONUS_DESTINY:
    case constants.BONUS_ENTANGLEMENT:
    case constants.BONUS_CHI:
    case constants.BONUS_CULTIVATION:
    case constants.BONUS_ONE_AMONG_N:
    case constants.BONUS_SKILL_RANK:
    case constants.BONUS_SPECIALITY:
      return true;
    case constants.BONUS_DISCOUNT:
    default:
      return false;
  }
}

export type TBonusesState = IBonus[];

export function createState(): TBonusesState { return []; }

export function applyBonuses(baseState: IStoreState, bonuses: IBonus[]): IStoreState {
  const nextState = produce(baseState, draftState => {
    bonuses.filter((bonus: IBonus) => isBonus(bonus))
    .forEach((bonus: IBonus) => {
      switch (bonus.type) {
        case constants.BONUS_DESTINY:
        draftState.destiny += bonus.value;
        break;
        case constants.BONUS_ENTANGLEMENT:
        draftState.entanglement += bonus.value;
        break;
        case constants.BONUS_CHI:
        chi.increase(draftState.chi, bonus.chi, bonus.value);
        break;
        case constants.BONUS_CULTIVATION:
        chi.increaseCultivation(draftState.chi, bonus.chi, bonus.value);
        break;
        case constants.BONUS_SKILL_RANK:
        const max = maxSkillBonus(draftState);
        skills.increase(draftState.skills, bonus.skill, max ? max : 0);
        break;
        case constants.BONUS_SPECIALITY:
        skills.addSpeciality(draftState.skills, bonus.skill, bonus.speciality);
        break;
        default:
        return;
      }
    });
  });
  return nextState;
}
