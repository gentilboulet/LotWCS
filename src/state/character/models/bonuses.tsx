import { ICharacterState } from "./type";

import { IBonus, isBonus } from "../../../perks/bonuses";
import * as constants from "../../../perks/constants/bonuses";

import * as chi from "./chi";

import { maxSkillBonus } from "./derived";
import * as skills from "./skills";

import * as loresheets from "./loresheets";

export type TBonusesState = IBonus[];

export function createState(): TBonusesState {
  return [];
}

export function applyBonuses(
  draftState: ICharacterState,
  bonuses: IBonus[]
): void {
  bonuses
    .filter((bonus: IBonus) => isBonus(bonus))
    .forEach((bonus: IBonus) => {
      switch (bonus.type) {
        case constants.BONUS_DESTINY:
          draftState.destiny += bonus.value;
          break;
        case constants.BONUS_ENTANGLEMENT:
          draftState.entanglement += bonus.value;
          break;
        case constants.BONUS_CHI:
          chi.increase(draftState.chi, bonus.chi, bonus.value);
          break;
        case constants.BONUS_CULTIVATION:
          chi.increaseCultivation(draftState.chi, bonus.chi, bonus.value);
          break;
        case constants.BONUS_SKILL_RANK:
          const max = maxSkillBonus(draftState);
          // TODO : overflows
          skills.increase(draftState.skills, bonus.skill, max);
          break;
        case constants.BONUS_SPECIALITY:
          // TODO : already speciality
          skills.addSpeciality(
            draftState.skills,
            bonus.skill,
            bonus.speciality
          );
          break;
        case constants.BONUS_LORESHEET:
          loresheets.openLoresheet(draftState.loresheets, bonus.lsUid);
          break;
        case constants.BONUS_LORESHEET_OPTION:
          loresheets.buyLoresheetOption(
            draftState.loresheets,
            bonus.lsUid,
            bonus.uid,
            bonus.payload
          );
          break;
        default:
          break;
      }
    });
}
