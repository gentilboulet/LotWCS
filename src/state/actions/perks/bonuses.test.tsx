import * as bonuses from './bonuses';

describe('Testing bonus creators', () => {
  it('should create a destiny bonus', () => {
    const b = bonuses.bonusDestiny(10);
    expect( b.value ).toBe(10);
    expect( b ).toMatchSnapshot();
  });

  it('should create an entanglement bonus', () => {
    const b = bonuses.bonusEntanglement(10);
    expect( b.value ).toBe(10);
    expect( b ).toMatchSnapshot();
  });

  it('should create an chi bonus', () => {
    const chiName = 'fire';
    const b = bonuses.bonusChi(10, chiName);
    expect( b.value ).toBe(10);
    expect( b.chi ).toBe(chiName);
    expect( b ).toMatchSnapshot();
  });

  it('should create an cultivation bonus', () => {
    const chiName = 'fire';
    const b = bonuses.bonusCultivation(chiName, 10);
    expect( b.value ).toBe(10);
    expect( b.chi ).toBe(chiName);
    expect( b ).toMatchSnapshot();
  });

  it('should create an oneAmongN bonus', () => {
    const bonusesList = [
      bonuses.bonusEntanglement(10),
      bonuses.bonusDestiny(10),
    ];

    const b = bonuses.bonusOneAmongN(bonusesList);
    expect( b ).toMatchSnapshot();
  });

  it('should not create an oneAmongN bonus with only one bonus but give the bonus back', () => {
    const bonusesList = [
      bonuses.bonusDestiny(10),
    ];

    const b = bonuses.bonusOneAmongN(bonusesList);
    expect( b ).toMatchObject(bonusesList[0]);
    expect( b ).toMatchSnapshot();
  })

  it('should not create an oneAmongN bonus with an empty set of bonuses', () =>{
    expect( () => bonuses.bonusOneAmongN([]) ).toThrow();
  });

  it('should create an skill rank bonus', () => {
    const skillName = 'Awareness';
    const b = bonuses.bonusSkillRank(skillName);
    expect( b.skill ).toBe(skillName);
    expect( b ).toMatchSnapshot();
  });

  it('should create a skill speciality bonus', () => {
    const skillName = 'Awareness';
    const specialityName = 'Hear';
    const b = bonuses.bonusSpeciality(skillName, specialityName);
    expect( b.skill ).toBe(skillName);
    expect( b.speciality ).toBe(specialityName);
    expect( b ).toMatchSnapshot();
  });


  it('should create a skill with a not in data speciality bonus', () => {
    const skillName = 'Awareness';
    const specialityName = 'Totally not in data speciality';
    const b = bonuses.bonusSpeciality(skillName, specialityName);
    expect( b.skill ).toBe(skillName);
    expect( b.speciality ).toBe(specialityName);
    expect( b ).toMatchSnapshot();
  });
});
