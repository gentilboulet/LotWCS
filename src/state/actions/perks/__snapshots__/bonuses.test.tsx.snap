// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing bonus creators should create a destiny bonus 1`] = `
Object {
  "type": "BONUS_DESTINY",
  "value": 10,
}
`;

exports[`Testing bonus creators should create a skill speciality bonus 1`] = `
Object {
  "skill": "Awareness",
  "speciality": "Hear",
  "type": "BONUS_SPECIALITY",
}
`;

exports[`Testing bonus creators should create a skill with a not in data speciality bonus 1`] = `
Object {
  "skill": "Awareness",
  "speciality": "Totally not in data speciality",
  "type": "BONUS_SPECIALITY",
}
`;

exports[`Testing bonus creators should create an chi bonus 1`] = `
Object {
  "chi": "fire",
  "type": "BONUS_CHI",
  "value": 10,
}
`;

exports[`Testing bonus creators should create an cultivation bonus 1`] = `
Object {
  "chi": "fire",
  "type": "BONUS_CULTIVATION",
  "value": 10,
}
`;

exports[`Testing bonus creators should create an entanglement bonus 1`] = `
Object {
  "type": "BONUS_ENTANGLEMENT",
  "value": 10,
}
`;

exports[`Testing bonus creators should create an oneAmongN bonus 1`] = `
Object {
  "bonuses": Array [
    Object {
      "type": "BONUS_ENTANGLEMENT",
      "value": 10,
    },
    Object {
      "type": "BONUS_DESTINY",
      "value": 10,
    },
  ],
  "type": "BONUS_ONE_AMONG_N",
}
`;

exports[`Testing bonus creators should create an skill rank bonus 1`] = `
Object {
  "skill": "Awareness",
  "type": "BONUS_SKILL_RANK",
}
`;

exports[`Testing bonus creators should not create an oneAmongN bonus with only one bonus but give the bonus back 1`] = `
Object {
  "type": "BONUS_DESTINY",
  "value": 10,
}
`;
